### PR TITLE
Add/shuffleboard

### PR DESCRIPTION
**Pull request by Tyson453**

### CODE REVIEW

File: robotbase/.vscode/launch.json [hunk 1]
```diff
@@ -4,18 +4,24 @@
   // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
   "version": "0.2.0",
   "configurations": [
```
Comment (NolanCampbell2357) [inline robotbase/.vscode/launch.json:6]: What caused this file to be modified? It should work the way it was.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/FeederSubsystem.java [hunk 2]
```diff
@@ -1,21 +1,21 @@
 package com.team2357.frc2022.subsystems;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/FeederSubsystem.java:1]: We should be using a functional interface, instead of a whole Arduino here. Also this change is not relevant to the current branch, the Arduino changes should have been done on their own branch.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package com.team2357.lib.arduino;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:1]: This should also have been done in it's own branch, as it is not relevant to the shuffleboard.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+public class ArduinoSensor extends ArduinoUSBController{
+
+    private String m_deviceName;
+
+    public ArduinoSensor(String ttyDevice, String deviceName) {
+        super(ttyDevice);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:8]: There is a bug where if the Arduino is not connected the robot code will crash via an IOException, this change should also find a way to fix that issue.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java (hunk 3 above)
Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:1]: Generally, we will only end up with one copy of this class, and will use it to fulfill functional interfaces for subsystems using a sensor.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package com.team2357.lib.arduino;
+
+public class ArduinoSensor extends ArduinoUSBController{
+
+    private String m_deviceName;
+
+    public ArduinoSensor(String ttyDevice, String deviceName) {
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:7]: We do not want to take a device name here, we want this class to handle all three sensors. Taking and using the device name makes your three member methods return the same value off the same object, instead of the actual sensor they are referring to.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        super(ttyDevice);
+        
+        m_deviceName = deviceName;
+    }
+
+    public boolean getIntakeValue() {
+        boolean value = false;
+        if (this.isConnected()){
+            value = !this.getDeviceFieldBoolean(m_deviceName, "state");
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:16]: Instead of m_deviceName we should be using a constant to refer to the specific device name.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        }
+        
+        return value;
+    }
+
+    public boolean getFeederValue() {
+        boolean value = false;
+        if (this.isConnected()){
+            value = !this.getDeviceFieldBoolean(m_deviceName, "state");
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:25]: Same here, replace m_deviceName with a constant. Usually we would not include constants in files like these, but this one is already implementation specific, so it does not hurt to directly use constants in here.

File: robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        }
+        
+        return value;
+    }
+
+    public boolean getTurretValue() {
+        boolean value = false;
+        if (this.isConnected()){
+            value = !this.getDeviceFieldBoolean(m_deviceName, "state");
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/arduino/ArduinoSensor.java:34]: Replace m_deviceName with constant here as well.

File: robotbase/src/main/java/com/team2357/lib/commands/setFailsafeCommand.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package com.team2357.lib.commands;
+
+// TODO: ShuffleBoard Implement
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/lib/commands/setFailsafeCommand.java:3]: This seems irrelevant with the changes made to the failSafeCommand, and would have the same functionality. This can be deleted.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/IntakeSubsystem.java [hunk 10]
```diff
@@ -45,4 +49,9 @@ public void setPivot(DoubleSolenoid.Value value) {
         m_intakeSolenoid.set(value);
     }
 
+    public boolean isBallAtSensor() {
+        boolean isSensorBlocked = false;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/IntakeSubsystem.java:53]: This method will go away once we have the functional interfaces.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/IntakeSubsystem.java [hunk 11]
```diff
@@ -16,13 +18,15 @@
...
     private VictorSPX m_intakeVictor;
+    private ArduinoSensor m_inputIRSensor;
 
     /**
      * @param intakeVictor Victor SPX to use to control intake
      */
     public IntakeSubsystem(VictorSPX intakeVictor, DoubleSolenoid intakeSolenoid) {
         m_intakeSolenoid = intakeSolenoid;
         m_intakeSolenoid.set(Value.kOff);
+        m_inputIRSensor = new ArduinoSensor(Constants.ARDUINO.ARDUINO_SENSOR_DEVICE_NAME, Constants.ARDUINO.IR_SENSOR_JSON_NAME);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/IntakeSubsystem.java:29]: This call will be irrelevant, but we should not be using constants in a subsystem. Subsystems are rarely purely implementation specific, so we would take these as either config or constructor parameters to make the code more reusable.

File: robotbase/src/main/java/com/team2357/frc2022/subsystems/FeederSubsystem.java [hunk 12]
```diff
@@ -28,10 +28,7 @@ public void runFeedermotor(double speed) {
     // flip that to true for reability
     public boolean isBallAtFeederWheel() {
         boolean isSensorBlocked = false;
-        if (m_arduinoIRSensor.isConnected()) {
-            isSensorBlocked = !m_arduinoIRSensor.getDeviceFieldBoolean(Constants.ARDUINO.IR_SENSOR_JSON_NAME,
-                    "state");
-        }
+        isSensorBlocked = m_outputIRSensor.getFeederValue();
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/subsystems/FeederSubsystem.java:31]: This method will also go away eventually, but could be simplified by just returning m_outputIRSensor.getFeederValue().

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
+package com.team2357.frc2022.shuffleboard;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:1]: This class will probably go away in favor of CargoNumberWidget, since it will require to less work to implement functional interface in it. However, the actual code and logic for the widget seems good.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        int numOfCargo = getNumOfCargo();
+        setNumOfCargo(++numOfCargo);
+        m_cargoCountWidget.setNumber(numOfCargo);
+        readyToSub = false;
+    }
+
+
+
+    public void periodic() {
+        if (readyToSub && !m_feederSub.isBallAtFeederWheel()) {
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:49]: We will have to check the internal size of the robot, but our robot may not be able to physically hold more than two cargo. That means the number of IR sensors that are blocked can determine the number of balls. We might be able to simplify this.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+            subCargo();
+        } else if (!readyToSub && m_feederSub.isBallAtFeederWheel()) {
+            readyToSub = true;
+        }
+        if (readyToAdd && !m_intakeSub.isBallAtSensor()) {
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:54]: I would consider renaming readyToAdd and readyToSub to with a name like ```intakeStateChange``` and ```feederStateChange```. This is just to make it a little clearer, since these booleans do not actually decide whether or not the robot can intake/shoot a cargo.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.wpilibj.shuffleboard.BuiltInWidgets;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+
+public class CargoCountWidget extends ShuffleboardWidget {
+    private final FeederSubsystem m_feederSub;
+    private final IntakeSubsystem m_intakeSub;
+    private static NetworkTableEntry m_cargoCountWidget;
+
+    private boolean readyToSub = false;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:15]: Remember that member variables should be prefixed with the "m_".

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        super(tabTitle);
+        m_feederSub = feederSub;
+        m_intakeSub = intakeSub;
+
+        NetworkTableEntry cargoCountWidget = Shuffleboard.getTab(tabTitle)
+            .add("Num of Cargo Balls", 0)
+            .withWidget(BuiltInWidgets.kTextView)
+            .getEntry();
+
+        m_cargoCountWidget = cargoCountWidget;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:29]: Unless the shuffleboard is weird, you should be able to do the ```Shuffleboard.getTab...``` directly on ```m_cargoCountWidget```.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoBallNumber.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
+package com.team2357.frc2022.shuffleboard;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoBallNumber.java:1]: This file needs renamed.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.wpilibj.shuffleboard.BuiltInWidgets;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+
+// TODO: ShuffleBoard Implement
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java:12]: ```TODO```s can be removed once you do whatever they say to do, same with any comments describing the issue.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    public static void show() {
+        Shuffleboard.selectTab((m_tabTitle));
+    }
+
+    public Command generateCommand() {
+        return new SequentialCommandGroup(
+        choosers[0].getWaitCommand(),
+        choosers[0].getActionCommand(),
+
+        choosers[0].getWaitCommand(),
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java:84]: We need to increment the ```choosers```. Currently each line is referring to the element at index 0, so we do not include commands in index 1, and 2. We want each element in the ```choosers``` to be included in the ```SequentialCommandGroup```.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]
We need to pull out the Arduino specific stuff into it's own branch to work on. It is irrelevant to the shuffleboard. Overall, the code looks good, and the logic seems sound. Tyson, I did not expect you to know how to do all this especially the stuff with the functional interfaces, so do not feel bad about your changes. The next time you are here I will help you with pulling the Arduino changes out and making the functional interfaces.

File: robotbase/.vscode/launch.json (hunk 1 above)
Comment (Tyson453) [reply robotbase/.vscode/launch.json:6]: I'm not sure why it was modified. I didn't mean to commit it.

File: robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java [hunk 21]
```diff
@@ -71,6 +78,20 @@ public RobotContainer() {
...
     m_driveSub.setDefaultCommand(new DriveProportionalCommand(m_driveSub, m_driverControls));
+
+    configureShuffleboard();
+  }
+
+  /**
+   * This method should set up the shuffleboard
+   */
+  public void configureShuffleboard() {
+    // TODO: ShuffleBoard Implement
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java:89]: You can remove these ```TODO```'s once you are finished.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.wpi.first.wpilibj2.command.WaitCommand;
+
+// TODO: ShuffleBoard Implement
+/**
+ * This class should control a chooser to select up to three actions with a wait
+ * command between them
+ * Do not worry about making the actual auto commands, just the chooser widget
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java:16]: Again, remove comments that you do not need.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+ * that can select different
+ * things is enough for now
+ */
+public class AutoModeCommandChooser {
+    public enum AutomodeActions {
+        NONE,
+        MOVE_OFF_LINE,
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java:23]: ```NONE``` is the only thing we know right now we will have right now. This means ```MOVE_OFF_LINE``` and ```SHOOT``` can go away.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+    private class AutoActionChooser {
+        protected NetworkTableEntry m_waitWidget;
+        protected SendableChooser<AutomodeActions> m_chooser;
+
+        protected AutoActionChooser(int index) {
+            ShuffleboardTab tab = Shuffleboard.getTab(m_tabTitle);
+            m_chooser = new SendableChooser<>();
+
+            m_chooser.setDefaultOption("None", AutomodeActions.NONE);
+            m_chooser.addOption("Shoot", AutomodeActions.SHOOT);
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/AutoModeCommandChooser.java:36]: The ```shoot``` and ```move off ``` options will go away here.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoNumberWidget.java [hunk 25]
```diff
@@ -0,0 +1,33 @@
+package com.team2357.frc2022.shuffleboard;
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoNumberWidget.java:1]: Merge this and CargoCountWidget and update to use the functional interfaces for the arduino sensors created in RobotContainer.java

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]

File: robotbase/.vscode/launch.json (hunk 1 above)
Comment (NolanCampbell2357) [reply robotbase/.vscode/launch.json:6]: It's still here.

File: robotbase/src/main/java/com/team2357/frc2022/commands/FailsafeCommand.java [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+        m_subsystem = subsystem;
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+
+        m_subsystem.setClosedLoopEnabled(m_failsafeActive);
+
+        CommandScheduler.getInstance().cancelAll();
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/commands/FailsafeCommand.java:23]: We should only cancel commands that require the subsystem being put in failsafe.

Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/frc2022/commands/FailsafeCommand.java:23]: Also, we should only cancel commands when the subsystem is being put into open loop. It is fine to keep running commands when going from open loop to closed loop.

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+
+
+
+    public void periodic() {
+        if (m_readyToSub && !m_feederSensor.getState()) {
+            subCargo();
+        } else if (!m_readyToSub && m_feederSensor.getState()) {
+            m_readyToSub = true;
+        }
+        if (m_readyToAdd && !m_intakeSensor.getState()) {
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:51]: We want to add cargo as soon as the sensor goes from false to true. We do not want to wait for the robot to move the cargo up to add.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]

File: robotbase/.vscode/launch.json (hunk 1 above)
Comment (Tyson453) [reply robotbase/.vscode/launch.json:6]: The most recent commit to this branch shows the same code in this file as the main code.

File: robotbase/src/main/java/com/team2357/lib/commands/setFailsafeCommand.java (hunk 9 above)
Comment (NolanCampbell2357) [reply robotbase/src/main/java/com/team2357/lib/commands/setFailsafeCommand.java:3]: Still need to delete this.

**Review by NolanCampbell2357 — CHANGES_REQUESTED** [review]
You are close

File: robotbase/src/main/java/com/team2357/frc2022/commands/FailsafeCommand.java [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+    public FailsafeCommand(boolean failsafeActive, ClosedLoopSubsystem subsystem) {
+        m_failsafeActive = failsafeActive;
+        m_subsystem = subsystem;
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+
+        m_subsystem.setClosedLoopEnabled(!m_failsafeActive);
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/commands/FailsafeCommand.java:21]: I think this code can be a little clearer. I'd suggest something like the following:
```
if failsafe active
  setClosedLoopEnabled(false)
  cancel commands
else
  setClosedLoopEnabled(true)
```

File: robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+    public void subCargo() {
+        int numOfCargo = getNumOfCargo();
+        setNumOfCargo(++numOfCargo);
+        m_cargoCountWidget.setNumber(numOfCargo);
+        m_feederStateChange = false;
+    }
+
+
+
+    public void periodic() {
```
Comment (coderkevin) [inline robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:45]: This periodic function is a bit confusing to me. I think it could be simplified and easier to read. @Tyson453 I'd be happy to pair with you on this.

Comment (coderkevin) [reply robotbase/src/main/java/com/team2357/frc2022/shuffleboard/CargoCountWidget.java:45]: Also, if we're going to go through the trouble of counting our cargo, I'm thinking we may want to use that elsewhere in the robot, so maybe there's a better place for these calculations.

**Review by coderkevin — COMMENTED** [review]
I'd suggest removing the CargoCountWidget, other than that this looks good, just one other suggestion

File: robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java [hunk 30]
```diff
@@ -70,6 +83,25 @@ public RobotContainer() {
...
+    configureShuffleboard();
+  }
+
+  /**
+   * This method should set up the shuffleboard
+   */
+  public void configureShuffleboard() {
+    DriveTab driveTab = new DriveTab();
+
+    driveTab.addWidget(new FailsafeButtonWidget(SHUFFLEBOARD_TAB_ROBOT, "Drive Failsafe", m_driveSub));
```
Comment (NolanCampbell2357) [inline robotbase/src/main/java/com/team2357/frc2022/RobotContainer.java:101]: You need to remove your references to deleted code. This is why the CI gradle build failed.

**Review by NolanCampbell2357 — APPROVED** [review]
